tests/performance/instrumentation_step_22.cc

```diff
@@ -0,0 +1,677 @@
+{
+  std::map<std::string, unsigned long long> cycle_count;
+
+  {
+    std::vector<unsigned int> subdivisions(dim, 1);
+    subdivisions[0] = 4;
+
+    const Point<dim> bottom_left = (dim == 2 ?                //
+                                      Point<dim>(-2, -1) :    // 2d case
+                                      Point<dim>(-2, 0, -1)); // 3d case
+
+    const Point<dim> top_right = (dim == 2 ?              //
+                                    Point<dim>(2, 0) :    // 2d case
+                                    Point<dim>(2, 1, 0)); // 3d case
+
+    GridGenerator::subdivided_hyper_rectangle(triangulation,
+                                              subdivisions,
+                                              bottom_left,
+                                              top_right);
+  }
+
+  for (const auto &cell : triangulation.active_cell_iterators())
+    for (const auto &face : cell->face_iterators())
+      if (face->center()[dim - 1] == 0)
+        face->set_all_boundary_ids(1);
+
+
+  switch (get_testing_environment())
+    {
+      case TestingEnvironment::light:
+        triangulation.refine_global(4 - dim);
+        break;
+      case TestingEnvironment::medium:
+        /* fallthrough */
+      case TestingEnvironment::heavy:
+        triangulation.refine_global(5 - dim);
+        break;
+    }
+
+  for (unsigned int refinement_cycle = 0; refinement_cycle < 6;
+       ++refinement_cycle)
+    {
+      debug_output << "Refinement cycle " << refinement_cycle << std::endl;
+
+      if (refinement_cycle > 0)
+        {
+          callgrind_wrapper::start_instrumentation();
+          refine_mesh();
+          cycle_count["refinement"] = callgrind_wrapper::stop_instrumentation();
+        }
+
+      callgrind_wrapper::start_instrumentation();
+      setup_dofs();
+      cycle_count["setup_system"] = callgrind_wrapper::stop_instrumentation();
+
+      debug_output << "   Assembling..." << std::endl << std::flush;
+      callgrind_wrapper::start_instrumentation();
+      assemble_system();
+      cycle_count["assemble_system"] =
+        callgrind_wrapper::stop_instrumentation();
+
+      callgrind_wrapper::start_instrumentation();
+      solve();
+      cycle_count["solve"] = callgrind_wrapper::stop_instrumentation();
+
+      callgrind_wrapper::start_instrumentation();
+      output_results(refinement_cycle);
+      cycle_count["output_results"] = callgrind_wrapper::stop_instrumentation();
+
+      debug_output << std::endl;
+    }
+  return {cycle_count["refinement"],
+          cycle_count["setup_system"],
+          cycle_count["assemble_system"],
+          cycle_count["solve"],
+          cycle_count["output_results"]};
+}
+
+
+std::tuple<Metric, unsigned int, std::vector<std::string>>
+describe_measurements()
+{
+  return {Metric::instruction_count,
+          1,
+          {"refinement",
+           "setup_system",
+           "assemble_system",
+           "solve",
+           "output_results"}};
```

Comment:
This design is not great, since it separates the numbers from their description. I don't want to hold the patch over this, but we should really reconsider this. I think that the function that does the timing should also report string labels for all of the numbers it returns. In other words, it should return the very `std::map` that you are building.

Reply:
I disagree on this point.

I think the design is perfectly acceptable for the purpose this header file serves [1].

The reason why I have split the description from the measurement is the fact that you need the information in `describe_measurement` once but you have to call `perform_single_measurement` multiple times.
I think this is cleaner than providing the names redudantly.

Now you can certainly argue that `perform_single_measurement` returning a map would be more elegant. But then you would need to write some logic sorting a `std::map<std::string, std::uint64_t>` return value into a `std::map<std::string, std::vector<std::uint64_t>>` (or similar) because you cannot guarantee that `perform_single_measurement` would always return a map with the same keys all the time.

[1] It is used internally for a grand total of four performance tests.